### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by e111077**
I have converted the project to typescript. Adding typescript people to take a look. Also, I haven't chatted at all with the team on this (kinda just did this on my own) so I'd like to have a discussion on whether this is a optimal solution to adding typescript support to pwa-helpers.

Reviewers added: TS nerds and project owners

I tried to follow best-practices on redux using this guide: https://github.com/piotrwitek/react-redux-typescript-guide

but I'm still not sure if I have missed some concept here or there.

### CODE REVIEW

**Comment by justinfagnani** [comment]
> I haven't chatted at all with the Kamino team on this

Seems like a critical thing to get agreement on. As much of a TypeScript fan as I am, all the maintainers should be fully on board before switching.

File: custom_typings/globals.d.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+interface Event {
+  composedPath(): EventTarget[]
+}
+
+interface EventInit {
+  composed?: boolean
+}
+
+interface HTMLElement {
+  connectedCallback(): void
```
Comment (rictic) [inline custom_typings/globals.d.ts]: These should be optional, i.e.: `connectedCallback(): void;`

**Comment by e111077** [comment]
I agree, it was kinda an exercise for myself to 1. understand Redux and 2. learn how TS interacts with it. I'm totally okay with rejecting this PR since it complicates the demo quite a bit, but we can also just take the outputted `.d.ts` files and add those to the project

File: package.json [hunk 2]
```diff
@@ -12,13 +12,18 @@
...
+    "typescript": "^2.8.3"
+  },
   "bugs": {
     "url": "https://github.com/polymer/pwa-helpers/issues"
   },
   "homepage": "https://github.com/polymer/pwa-helpers#readme",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "build": "tsc"
```
Comment (rictic) [inline package.json]: We should build on travis, even if we don't have tests.

File: src/axe-report.ts [hunk 3]
```diff
@@ -38,14 +38,33 @@ subject to an additional IP rights grant found at http://polymer.github.io/PATEN
     });
   });
 */
```
Comment (rictic) [inline src/axe-report.ts]: Here and elsewhere, this comment should stay attached to the function, for the sake of human & machine :)

File: src/connect-mixin.ts [hunk 4]
```diff
@@ -35,15 +37,15 @@ export const connect = (store) => (baseElement) => class extends baseElement {
...
 
     if (super.disconnectedCallback) {
       super.disconnectedCallback();
     }
   }
 
   // This is called every time something is updated in the store.
-  _stateChanged(state) {
-    throw new Error('_stateChanged() not implemented', this);
+  _stateChanged(_state: S) {
```
Comment (rictic) [inline src/connect-mixin.ts]: This looks `protected`

File: src/axe-report.ts [hunk 5]
```diff
@@ -38,14 +38,33 @@ subject to an additional IP rights grant found at http://polymer.github.io/PATEN
...
+  }
+}
+
+export interface AxeReportOptions extends axeTypes.RunOptions {
+  cleanup?: () => Promise<void>
+  axeConfig?: axeTypes.RunOptions
+}
+
+interface AxeReportRunOptions extends axeTypes.RunOptions {
+  resultTypes: string[]
```
Comment (rictic) [inline src/axe-report.ts]: This is missing from the upstream type? Consider submitting a PR with this and the other missing pieces to DefinitelyTyped

Comment (rictic) [inline src/axe-report.ts]: This looks like it should be optional

File: src/axe-report.ts [hunk 6]
```diff
@@ -38,14 +38,33 @@ subject to an additional IP rights grant found at http://polymer.github.io/PATEN
...
   const {violations} = await axe.run(dom, axeConfig || {
-    runOnly: ['wcag2a', 'wcag2aa', 'section508'],
+    runOnly: {
+      type: 'tag',
+      values: ['wcag2a', 'wcag2aa', 'section508']
+    },
     // Ignore tests that are passing.
     resultTypes: ['violations']
-  });
+  } as AxeReportRunOptions);
```
Comment (rictic) [inline src/axe-report.ts]: `as` should be used as little as possible, as it's more powerful than you usually mean. One way to write this:

```typescript
// At the top level
const defaultConfig: AxeReportRunOptions = {
  runOnly: {
    type: 'tag',
    values: ['wcag2a', 'wcag2aa', 'section508']
  },
  // Ignore tests that are passing.
  resultTypes: ['violations']
};

// then in axeReport
const {violations} = await axe.run(dom, axeConfig || defaultConfig);
```

File: src/demo/actions/counter.ts [hunk 7]
```diff
@@ -8,17 +8,26 @@ Code distributed by Google as part of the polymer project is also
...
 export const INCREMENT = 'INCREMENT';
 export const DECREMENT = 'DECREMENT';
 
+export interface CounterActionIncrement extends r.Action<'INCREMENT'> {};
+export interface CounterActionDecrement extends r.Action<'DECREMENT'> {};
+export type CounterAction = CounterActionIncrement | CounterActionDecrement;
+
 export const increment = () => {
-  return {
+  const action: CounterActionIncrement = {
```
Comment (rictic) [inline src/demo/actions/counter.ts]: Style note, you can get the same result without introducing a local by annotating the function with the return type. i.e. `export const increment = (): CounterActionIncrement => {`

File: src/demo/components/counter-element.ts [hunk 8]
```diff
@@ -56,9 +56,12 @@ class CounterElement extends HTMLElement {
   }
 
   _show() {
-    this.shadowRoot.getElementById('clicksSpan').textContent = this.clicks;
-    this.shadowRoot.getElementById('valueSpan').textContent = this.value;
+    this.shadowRoot!.getElementById('clicksSpan')!.textContent = this.clicks.toString();
```
Comment (rictic) [inline src/demo/components/counter-element.ts]: Can avoid the need for the exclamation points by declaring that the shadow root is a property that is not undefined. i.e. writing:
```typescript
  shadowRoot!: ShadowRoot;
```

with the other class properties

File: src/demo/components/redux-example.ts [hunk 9]
```diff
@@ -37,39 +39,46 @@ template.innerHTML = `
 `
 
 class ReduxExample extends connect(store)(HTMLElement) {
+  _counter: CounterElement
```
Comment (rictic) [inline src/demo/components/redux-example.ts]: protected? private?

File: src/demo/components/redux-example.ts [hunk 10]
```diff
@@ -37,39 +39,46 @@ template.innerHTML = `
...
+    this._clicksSpan.textContent = numClicks.toString();
+    this._valueSpan.textContent = value.toString();
+    this._didLoadSpan.textContent = state.lazy ? state.lazy.didLoad.toString() : 'undefined';
   }
 
   _loadReducer() {
     import('../reducers/lazy.js').then((module) => {
       const reducer = module.default;
-      store.addReducers({'lazy': reducer});
+      const reducerMap: r.ReducersMapObject<AppStateLazy> = {'lazy': reducer};
```
Comment (rictic) [inline src/demo/components/redux-example.ts]: This more specific type is nice I guess, but is it necessary?

I suppose it is additional documentation for what you're trying to do, which is nice.

File: src/demo/components/redux-example.ts [hunk 11]
```diff
@@ -37,39 +39,46 @@ template.innerHTML = `
...
-    this._clicksSpan.textContent = this._counter.clicks = state.counter.clicks;
-    this._valueSpan.textContent = this._counter.value = state.counter.value;
-    this._didLoadSpan.textContent = state.lazy ? state.lazy.didLoad : 'undefined';
+    this._clicksSpan.textContent = numClicks.toString();
+    this._valueSpan.textContent = value.toString();
+    this._didLoadSpan.textContent = state.lazy ? state.lazy.didLoad.toString() : 'undefined';
   }
 
   _loadReducer() {
     import('../reducers/lazy.js').then((module) => {
```
Comment (rictic) [inline src/demo/components/redux-example.ts]: it's a little weird that this promise isn't awaited or returned. is this a race?

File: src/router.ts [hunk 12]
```diff
@@ -37,13 +37,12 @@ subject to an additional IP rights grant found at http://polymer.github.io/PATEN
...
 */
-
-export const installRouter = (locationUpdatedCallback) => {
+export const installRouter = (locationUpdatedCallback: (location:Location, event: Event|null) => void) => {
   document.body.addEventListener('click', e => {
     if (e.defaultPrevented || e.button !== 0 ||
         e.metaKey || e.ctrlKey || e.shiftKey) return;
 
-    const anchor = e.composedPath().filter(n => n.tagName === 'A')[0];
+    const anchor = e.composedPath().filter(n => n instanceof HTMLAnchorElement && n.tagName === 'A')[0] as HTMLAnchorElement;
```
Comment (rictic) [inline src/router.ts]: `array.filter(p)[0]` can be rewritten as `array.find(p)`

Comment (rictic) [inline src/router.ts]: I think you want `as HTMLAnchorElement | undefined`

File: src/connect-mixin.ts (hunk 4 above)
Comment (e111077) [reply src/connect-mixin.ts]: looks like this is a TS no-go: https://github.com/Microsoft/TypeScript/issues/17293

File: src/axe-report.ts (hunk 5 above)
Comment (e111077) [reply src/axe-report.ts]: this is correct. My PR got merged today looks like it has not been released:

https://github.com/dequelabs/axe-core/pull/568

**Comment by merlinnot** [comment]
If you want to have some user voice on this... It would be great :) But if you decide to keep it plane js, we can always publish `@types/pwa-helpers`, so it's not a must-have :)

**Comment by e111077** [comment]
Well, we now have correct types that can be used!

File: src/axe-report.ts [hunk 13]
```diff
@@ -38,6 +38,7 @@ subject to an additional IP rights grant found at http://polymer.github.io/PATEN
     });
   });
 */
+// named differently from axe so that the library is not pulled in w/compilation
```
[the remark this answers is not in the document]
Comment (e111077) [reply src/axe-report.ts]: WOW, that is some ugly syntax

**Comment by notwaldorf** [comment]
Hiiiiii ok so we did some offline chatting about this and it goes like this:
- we like types! we support adding types!
- at the moment this PR has no dependencies and no tool requirements, which makes it really nice and easy to use eeeeeeverywhere. We'd like to keep this as a feature
- as a result, we'd like to keep the source in js (so we would also not have to deal with building and praying the types away in the build-published-to-npm version
- based on this discussion (i know nothing about TS tools), i understand we have 2 options: publish the `.ts` typings file to this repo, or publishing them to `@types`.

We talked offline and your preference between these two seems to be for `@types`, so uhhh let's do that?

**Comment by merlinnot** [comment]
There is one difference: if we have codebase in TypeScript, our types are always up to date and we have (at least people who like TypeScript) better development experience.

We can add a prepublish hook to build it and we can easily get rid of the source if we put it into one directory (e.g. `src`) and add `files` to the `package.json`.

To be super safe, we can also add a precommit hook so no one commits stuff which is not compiling or does not pass tslint.

TypeScript is not only about consumers of this package, but also about people developing it.

I consider `@types` as the last resort.

@notwaldorf would you guys be more willing to merge this if I would improve this PR by adding a godlike config and a readme for people less experienced with TypeScript?

**Comment by notwaldorf** [comment]
@merlinnot this is the option 2 i listed, right? ("publish the `.ts` files to this repo). I asked the local TS experts, @e111077 and @rictic and they both suggested `@types`, so i defer to their decision.

I don't understand from your comment what "godlike config means", and whether this still means changing the source to TS -- this is something we are not planning on doing at this time.

**Comment by merlinnot** [comment]
If you are not planning to do that, than `@types` are indeed a better option. I was referring to your last sentence, "preference between these two seems to be for @types".

If we'd have a source in TS, than we wouldn't get out of sync, it happens a lot with `@types`.

For a config I mean something that is pleasant to work with and does not allow anyone to mess up :)

**Comment by notwaldorf** [comment]
Going to close this PR since my understanding is that the change isn't needed here (and it's needed in the `@types` repo instead). Pls reopen if I misunderstood our conclusion!

**Pull request closed without merging**